Support for pinning tables at top of tables list #added

ShashwatC opened the pull request:
## Changes:
- Added "pin table" menu items in DBView UI for gear menu and context menu.
- Supporting pin table functionality with and without table filtering.

## Closes following issues:
- Closes: https://github.com/Sequel-Ace/Sequel-Ace/issues/1225

## Tested:
- Processors:
  - [x] Intel
  - [ ] Apple Silicon
- macOS Versions:
  - [ ] 10.12.x (Sierra)
  - [ ] 10.13.x (High Sierra)
  - [ ] 10.14.x (Mojave)
  - [ ] 10.15.x (Catalina)
  - [x] 11.x (Big Sur)
- Localizations:
  - [x] English
  - [ ] Spanish
  - [ ] Other (please specify)
- Xcode Version: 13.0

Kaspik commented on Resources/Localization/zh-Hant.lproj/Localizable.strings:
```diff
@@ -3311,3 +3311,23 @@
 /* Info alert when editing a generated column */
 "The current field \"%@\" is a generated column and therefore cannot be edited." = "The current field \"%@\" is a generated column and therefore cannot be edited.";
+
+/* Pin a view to the top of the list of tables
```
Next time these can be added only to `en` strings and build will take care of translating and copying to all other languages :)

ShashwatC replied:
Got it. Am using JetBrains AppCode instead of xcode because of my familiarity with intelliJ IDEA and all the IDE features and shortcuts. The IDE created the strings in all language files for me so I let it be that way.

Jason-Morcos commented:
This is some awesome work! I feel right now though that it's not very clear that pinned tables aren't persisted/remembered. Either there should probably be some clarity added around this ("Temporarily Pin..." or something) or persistence should be added (the better option in my opinion). I think without doing one of the previous two things, we'll get lots of issues with people not understanding the feature and thinking it's "broken because it's not remembering the tables".

ShashwatC commented:
Thanks @Jason-Morcos , I plan to add support for persistence as well as the ability to unpin tables. Wanted to get one round of review in first, so I can learn if I'm implementing anything in a non-idiomatic way, going contrary to the larger design, etc., I can refactor now rather than doing it later when the PR is much bigger.

Also, for implementing unpin functionality, wanted to get your opinions on which approach seems better-
1. Two distinct menu items, with 2 different corresponding outlets and actions - one for pinning, one for unpin
2. Single menu item, with one outlet and one action - that menu item will call "toggle pin" function, and the title of that item will be pin or unpin depending on whether table is pinned.

Visagalis commented:
> Thanks @Jason-Morcos , I plan to add support for persistence as well as the ability to unpin tables. Wanted to get one round of review in first, so I can learn if I'm implementing anything in a non-idiomatic way, going contrary to the larger design, etc., I can refactor now rather than doing it later when the PR is much bigger.
> 
> Also, for implementing unpin functionality, wanted to get your opinions on which approach seems better-
> 
> 1. Two distinct menu items, with 2 different corresponding outlets and actions - one for pinning, one for unpin. Only one of the items will be visible at any time depending on whether table is pinned.
> 2. Single menu item, with one outlet and one action - that menu item will call "toggle pin" function, and the title of that item will be pin or unpin depending on whether table is pinned.

I would go for 2nd option.

Visagalis commented:
> Had few questions related to persistence-
> 
> 1. Thinking of doing something similar to SQLiteHistoryManager for persisting pinned tables. Wanted to check if some other way of implementing it would be preferred, or persisting with sqlite works.
> 2. Since pinned tables need to be maintained at a database level, wanted to know if there is already something in place in sequel ace to uniquely represent a database, or should I create some unique identifier myself. A (hostname, database name) tuple should work.

@Jason-Morcos maybe you could help with these questions related to persistence?

ShashwatC commented:
@Jason-Morcos Thanks for the suggestions. I have implemented persistence with SQLite, and pinning/unpinning table is getting synced across multiple tabs using NSNotifications. There are still a few kinks to work out, but overall things are working well. Can you go through the code and let me know if you have any comments?

Jason-Morcos commented on Resources/Localization/en.lproj/Localizable.strings:
```diff
@@ -3311,3 +3311,43 @@
+
+/* Pin a view to the top of the list of tables
+ Table List : Context Menu : Pin View
+ Table List : Gear Menu : Pin View */
+"Pin View" = "Pin View";
+
+/* Pin a table to the top of the list of tables
+ Table List : Context Menu : Pin Table
+ Table List : Gear Menu : Pin Table */
+"Pin Table" = "Pin Table";
+
+/* Pin a procudure to the top of the list of tables
+ Table List : Context Menu : Pin Procedure
+ Table List : Gear Menu : Pin Procedure */
+"Pin Procedure" = "Pin Procedure";
+
+/* Pin a function to the top of the list of tables
+ Table List : Context Menu : Pin Function
+ Table List : Gear Menu : Pin Function */
+"Pin Function" = "Pin Function";
+
+/* Unpin a view that is currently pinned at the top of the list of tables
+ Table List : Context Menu : Unpin View
+ Table List : Gear Menu : Unpin View */
+"Unpin View" = "Unpin View";
+
+/* Unpin a table that is currently pinned at the top of the list of tables
+ Table List : Context Menu : Unpin Table
+ Table List : Gear Menu : Unpin Table */
+"Unpin Table" = "Unpin Table";
+
+/* Unpin a procudure that is currently pinned at the top of the list of tables
+ Table List : Context Menu : Unpin Procedure
+ Table List : Gear Menu : Unpin Procedure */
+"Unpin Procedure" = "Unpin Procedure";
+
+/* Unpin a function that is currently pinned at the top of the list of tables
+ Table List : Context Menu : Unpin Function
+ Table List : Gear Menu : Unpin Function */
+"Unpin Function" = "Unpin Function";
```
No need to add the language strings to all the other languages. I would go ahead and remove them from all the other languages and just set them up for English - we use crowdin to do localization and it will just re-create all the other language files filling in missing ones based on the English one. Keeps things clearer if you only edit the one (English) lang file

Jason-Morcos commented on Source/Controllers/SubviewControllers/SPTablesList.m:
```diff
@@ -2192,6 +2407,8 @@ - (void)_removeTable:(BOOL)force
 		[filteredTableTypes removeLastObject];
 	}
 
+	// TODO : Remove the pinned item if appropriate
```
Just flagging the TODO

Jason-Morcos replied:
Still flagging todo, but I think it just now needs to be removed?

Jason-Morcos commented on Source/Controllers/SubviewControllers/SPTablesList.m:
```diff
@@ -1062,6 +1148,13 @@ - (void)setSelectionState:(NSDictionary *)selectionDetails
 		[openTableInNewTabContextMenuItem setTitle:NSLocalizedString(@"Open Table in New Tab", @"open table in new tab title")];
 		[openTableInNewWindowContextMenuItem setTitle:NSLocalizedString(@"Open Table in New Window", @"Table List : Context Menu : Duplicate connection to new window")];
+        [pinTableContextMenuItem setHidden:NO];
+        if (isCurrentSelectionPinned) {
+            [pinTableContextMenuItem setTitle:NSLocalizedString(@"Unpin Table", @"pin view menu item title")];
+        }
+        else {
+            [pinTableContextMenuItem setTitle:NSLocalizedString(@"Pin Table", @"pin view menu item title")];
+        }
```
Should the description text say table instead of view?

Jason-Morcos commented on Source/Controllers/SubviewControllers/SPTablesList.m:
```diff
@@ -1044,6 +1123,13 @@ - (void)setSelectionState:(NSDictionary *)selectionDetails
 		[openTableInNewTabMenuItem setTitle:NSLocalizedString(@"Open Table in New Tab", @"open table in new table title")];
 		[openTableInNewWindowMenuItem setTitle:NSLocalizedString(@"Open Table in New Window", @"Table List : Gear Menu : Duplicate connection to new window")];
+		[pinTableMenuItem setHidden:NO];
+        if (isCurrentSelectionPinned) {
+            [pinTableMenuItem setTitle:NSLocalizedString(@"Unpin Table", @"pin view menu item title")];
+        }
+        else {
+            [pinTableMenuItem setTitle:NSLocalizedString(@"Pin Table", @"pin view menu item title")];
+        }
```
Should the description text say table instead of view?

Jason-Morcos commented on Source/Controllers/SubviewControllers/SPTablesList.m:
```diff
@@ -1103,6 +1203,13 @@ - (void)setSelectionState:(NSDictionary *)selectionDetails
 		[openTableInNewTabContextMenuItem setHidden:NO];
 		[openTableInNewWindowContextMenuItem setHidden:NO];
+        [pinTableContextMenuItem setHidden:NO];
+        if (isCurrentSelectionPinned) {
+            [pinTableContextMenuItem setTitle:NSLocalizedString(@"Unpin Procedure", @"pin view menu item title")];
+        }
+        else {
+            [pinTableContextMenuItem setTitle:NSLocalizedString(@"Pin Procedure", @"pin view menu item title")];
+        }
```
Should the description text say procedure instead of view?

Jason-Morcos commented on Source/Controllers/SubviewControllers/SPTablesList.m:
```diff
@@ -1134,6 +1241,13 @@ - (void)setSelectionState:(NSDictionary *)selectionDetails
 		[separatorTableMenuItem3 setHidden:NO];
 		[openTableInNewTabMenuItem setTitle:NSLocalizedString(@"Open Function in New Tab", @"open function in new table title")];
 		[openTableInNewWindowMenuItem setTitle:NSLocalizedString(@"Open Function in New Window", @"Table List : Gear Menu : duplicate connection to new window")];
+        [pinTableMenuItem setHidden:NO];
+        if (isCurrentSelectionPinned) {
+            [pinTableMenuItem setTitle:NSLocalizedString(@"Unpin Function", @"pin view menu item title")];
+        }
+        else {
+            [pinTableMenuItem setTitle:NSLocalizedString(@"Pin Function", @"pin view menu item title")];
+        }
```
Should the description text say function instead of view?

Jason-Morcos commented on Source/Controllers/SubviewControllers/SPTablesList.m:
```diff
@@ -1150,6 +1264,13 @@ - (void)setSelectionState:(NSDictionary *)selectionDetails
 		[separatorTableContextMenuItem3 setHidden:NO];
 		[openTableInNewTabContextMenuItem setTitle:NSLocalizedString(@"Open Function in New Tab", @"open function in new table title")];
 		[openTableInNewWindowContextMenuItem setTitle:NSLocalizedString(@"Open Function in New Window", @"Table List : Context Menu : duplicate connection to new window")];
+        [pinTableContextMenuItem setHidden:NO];
+        if (isCurrentSelectionPinned) {
+            [pinTableContextMenuItem setTitle:NSLocalizedString(@"Unpin Function", @"pin view menu item title")];
+        }
+        else {
+            [pinTableContextMenuItem setTitle:NSLocalizedString(@"Pin Function", @"pin view menu item title")];
+        }
```
Should the description text say function instead of view?

Jason-Morcos commented on Source/Controllers/SubviewControllers/SPTablesList.m:
```diff
@@ -994,6 +1059,13 @@ - (void)setSelectionState:(NSDictionary *)selectionDetails
 		[separatorTableMenuItem3 setHidden:NO];
 		[openTableInNewTabMenuItem setTitle:NSLocalizedString(@"Open View in New Tab", @"open view in new table title")];
 		[openTableInNewWindowMenuItem setTitle:NSLocalizedString(@"Open View in New Window", @"Tables List : Gear Menu : Duplicate connection to new window")];
+		[pinTableMenuItem setHidden:NO];
+        if (isCurrentSelectionPinned) {
+            [pinTableMenuItem setTitle:NSLocalizedString(@"Unpin View", @"pin view menu item title")];
+        }
+        else {
+            [pinTableMenuItem setTitle:NSLocalizedString(@"Pin View", @"pin view menu item title")];
+        }
```
For all of these text spots, it looks like you're getting the text twice (like the if block is exactly duplicated)? Maybe you should get it once and store it in a variable or abstract it to a function?

Jason-Morcos commented on Source/Controllers/SubviewControllers/SPTablesList.m:
```diff
@@ -1937,6 +2058,83 @@ - (void) clearFilter
+    NSString * databaseName = [mySQLConnection database];
+    NSArray *tablesToPin = [_SQLitePinnedTableManager getPinnedTablesWithHostName:hostName databaseName:databaseName];
+    if (tablesToPin.count > 0) {
+        [tables insertObject:@"PINNED" atIndex:0];
```
What happens here if the _previous_ tablesToPin count was greater than zero and the PINNED object was already inserted? Maybe instead do a if tablesToPin.count > 0 and safeObjectAtIndex(0) (maybe function name wrong, something safeObject... away from Xcode right now) != "PINNED" then insert?
In other words, check if index 0 is already PINNED before inserting.
Also, does that phrase "PINNED" need to be localized in the few places its handled? Is table/view/procedure localized?

Jason-Morcos replied:
I guess in general here, it may be worth making sure there aren't pinned tables already and handling cleanup if they are? Or is this function guaranteed to only be called once ever?

Jason-Morcos commented on Source/Controllers/SubviewControllers/SPTablesList.m:
```diff
@@ -1937,6 +2058,83 @@ - (void) clearFilter
+}
+
+
+- (void)usePinTableNotification:(nonnull NSNotification *) notification; {
```
This is pretty cool and elegant actually, but I'd maybe rather think about trying to handle adds and removes by making the initPinnedTables function re-run safe and just reloading the pinned tables anytime there's a change in another db view? That might make the logic in initPinnedTables a little more complex, but would mean there aren't two chunks of code messing with the pinned tables section?

Jason-Morcos reviewed (commented):
More great progress on this one! These are the comments I have

Visagalis commented:
@ShashwatC Hello buddy, how it's going? Hope Jason didn't demotivate you on this feature?
I'm looking forward to it!

ShashwatC commented:
@Visagalis With remote work ending, I was a bit busy with relocating to another city. Would pick up this feature again in few days, now that I'm settled in.

Jason-Morcos commented:
Looks like some good progress here, @ShashwatC, but looks like the test cases are failing! Let me know if you need any help on this one, otherwise lmk when you've got the tests passing and I'll take a look the the PR! :)

ShashwatC commented:
Hey @Jason-Morcos, I had missed pushing project.pbxproj changes and build was failing, can you trigger the run tests workflow. It should work now.

Jason-Morcos commented on Resources/Localization/en.lproj/Localizable.strings:
```diff
@@ -3317,3 +3317,43 @@
 /* Info alert when editing a generated column */
 "The current field \"%@\" is a generated column and therefore cannot be edited." = "The current field \"%@\" is a generated column and therefore cannot be edited.";
+
+/* Pin a view to the top of the list of tables
+ Table List : Context Menu : Pin View
+ Table List : Gear Menu : Pin View */
```
Maybe worth making this comment a single line and a little more direct like the other localization string descriptions are
ex. `/* Menu option to pin a view to the top of the list of tables */`

Jason-Morcos commented on Source/Controllers/Other/SQLitePinnedTableManager.swift:
```diff
@@ -0,0 +1,202 @@
+
+    @objc static let sharedInstance = SQLitePinnedTableManager()
+    @objc private var queue: FMDatabaseQueue
+    private let Log = OSLog(subsystem: "com.sequel-ace.sequel-ace", category: "queryDatabase")
```
Replace `queryDatabase` with something else (ex. `pinnedTablesDatabase`)

Jason-Morcos commented on Source/Controllers/SubviewControllers/SPTablesList.m:
```diff
@@ -1937,6 +1991,88 @@ - (void) clearFilter
+ */
+- (void)removePinnedTablesSection {
+    [pinnedTables removeAllObjects];
+    NSString * pinnedHeader = NSLocalizedString(@"PINNED", @"header for pinned tables");
```
`NSLocalizedString(@"PINNED", @"header for pinned tables");`
I believe this string is missing from the localization file!

Jason-Morcos reviewed (changes requested):
Looking great so far! I was able to get it to crash though and found a couple weird things:
- If you create a new table after at least one table is pinned, that new table is added to the "pinned" section in the UI even though the table isn't actually pinned
- If you open the same database connection in two tabs and then create a new table and pin it in one of those tabs, everything crashes - I think the issue is that creating a table in one tab doesn't cause the other tabs to reload their table lists, just their _pinned_ tables - so it tries to "pin" a table that isn't in the list yet.

ShashwatC commented:
Have fixed the bugs in 'create table' codepath-

- New table getting created in pinned section because of how index of insertion is decided based on the alphabetical order of table names. 
- App crash due to trying to pin table which doesn't exist in that tab.

Also made some changes in 'delete table' flow. If someone deletes the table by selecting the pinned item, then the table name sti. l remains in the unpinned list. I have triggered a full reload on table deletion to fix that issue. Not sure if its a good idea, I can remove the problematic tables with more granular code finding and removal too if that's better.

Jason-Morcos reviewed (changes requested):
This is looking AWESOME! Great work!

I found another little thing - renaming tables.
Renaming tables that are pinned has issues if you rename them from the pinned list or from the main list both. Renaming them in the main list causes them to not be renamed in the pinned list and if you click on the table in the pinned list you then get an error. Renaming them in the pinned list does the opposite - fixes them in the pinned list but not in the main list.

I think it would be fine to do the same kind of table list reload after renames. A more granular approach would probably be better, but it's probably simpler to just reload after the rename. We will have to make sure that the rename updates the name of the entry in the pinned tables list, however! :)

Jason-Morcos commented:
More of a discussion thing in general for those following along - should pinned tables be sorted alphabetically or in order of when they were pinned?

samuelbradshaw commented:
> More of a discussion thing in general for those following along - should pinned tables be sorted alphabetically or in order of when they were pinned?

I think the default sort order should depend on if the user is able to manually reorder them.
- If a user is able to manually reorder the pinned tables list, I would expect newly-pinned tables to be added to the bottom of the pinned tables list.
- If a user is not able to manually reorder the pinned tables list, I would expect pinned tables to be listed alphabetically.

Jason-Morcos commented:
> > More of a discussion thing in general for those following along - should pinned tables be sorted alphabetically or in order of when they were pinned?
> 
> I think the default sort order should depend on if the user is able to manually reorder them.
> 
> * If a user is able to manually reorder the pinned tables list, I would expect newly-pinned tables to be added to the bottom of the pinned tables list, with the expectation that the user will then move them into their preferred position.
> * If a user is not able to manually reorder the pinned tables list, I would expect pinned tables to be listed alphabetically.

Currently the list is not reorderable and new pinned items are added to the top.
I think alphabetizing the pinned items is probably a good idea, especially to your point @samuelbradshaw if they're not re-orderable

Jason-Morcos commented:
@ShashwatC The work you've done here is awesome and pretty close to done I think! I'd love to include it in the next release, do you have an estimate for when you'd be able to make the last couple tweaks to get this ready to go?

ShashwatC commented:
Hey @Jason-Morcos, pushed the following changes

- fix for a small bug in the delete workflow.
- when renaming a pinned table - unpin the original name, pin the new name, take care of everything else via a refresh.
- for alphabetical order I'm sorting the pinned tables, then adding to the tables array in the required order.
- added unique key constraint to SQLite table, and other checks to ensure a single table doesn't get pinned multiple times in any circumstance.

The feature is about done, I'll incorporate any further comments quickly and get it included in next release 😀

Jason-Morcos commented:
Awesome! This looks great! Could you pull back changes from latest main and resolve the conflict with the project.pbxproj file? Then we'll give it a final look and merge!

ShashwatC commented:
Resolved the merge conflict @Jason-Morcos , please take a look.

Jason-Morcos reviewed (approved):
This looks great and will be a wonderful feature for users! Great work, @ShashwatC!

Kaspik commented on Source/Controllers/Other/SQLitePinnedTableManager.swift:
```diff
@@ -130,6 +131,10 @@ import OSLog
 
     @objc func pinTable(hostName: String, databaseName: String, tableToPin: String) {
+        
+        if (pinnedTablesDatabaseDictionary[hostName]?[databaseName] != nil && pinnedTablesDatabaseDictionary[hostName]![databaseName]!.contains(tableToPin)) {
```
`if let database = pinnedTablesDatabaseDictionary[hostName]?[databaseName], database.contains(tableToPin)`

Let's try to avoid force unwrap

Kaspik commented on Source/Controllers/Other/SQLitePinnedTableManager.swift:
```diff
@@ -0,0 +1,207 @@
+        if pinnedTablesDatabaseDictionary[hostName]?[databaseName] == nil {
+            return
+        }
+        if !pinnedTablesDatabaseDictionary[hostName]![databaseName]!.contains(tableToUnpin) {
```
^^ - try to avoid force unwrap via getting the `database` to `let` variable instead of check for null

```swift
guard let database = pinnedTablesDatabaseDictionary[hostName]?[databaseName], database.contains(tableToUnpin) else {
   return
}
```

Kaspik commented on Source/Controllers/Other/SQLitePinnedTableManager.swift:
```diff
@@ -0,0 +1,207 @@
+            return
+        }
+        let index = pinnedTablesDatabaseDictionary[hostName]![databaseName]!.firstIndex(of: tableToUnpin)!
+        pinnedTablesDatabaseDictionary[hostName]![databaseName]!.remove(at: index)
```
`database.remove(at: index)`

Kaspik replied:
btw. isn't there method to `removeAll(where: { $0 == tableToUnpin })` so u don't even need to get index etc?

Kaspik commented on Source/Controllers/Other/SQLitePinnedTableManager.swift:
```diff
@@ -0,0 +1,207 @@
+        if pinnedTablesDatabaseDictionary[hostName] == nil {
+            pinnedTablesDatabaseDictionary[hostName] = [:];
+        }
+        if pinnedTablesDatabaseDictionary[hostName]?[databaseName] == nil {
```
^^ can we work with local var instead of accessing all this all the time?

ShashwatC replied:
By working with local variable, you mean something like this?
`if pinnedTablesDatabaseDictionary[hostName] == nil {
            pinnedTablesDatabaseDictionary[hostName] = [:];
        }
        var pinnedTablesForHost = pinnedTablesDatabaseDictionary[hostName];
        if pinnedTablesForHost?[databaseName] == nil {
            pinnedTablesForHost?[databaseName] = []
        }
        var pinnedTablesForHostAndDatabase = pinnedTablesForHost?[databaseName];
        pinnedTablesForHostAndDatabase?.append(tableToPin)`

Kaspik reviewed (commented):
Just a couple of small Swift improvements. Please try to avoid force unwrap as much as possible :)

ShashwatC commented:
Hey, I was testing out my latest changes, looks like some bugs are there having to do with me misinterpreting some pass-by value/reference stuff. Will push a fix in sometime.

ShashwatC commented:
I have fixed the issues. 
```swift
if var pinnedTables = pinnedTablesDatabaseDictionary[hostName]?[databaseName], pinnedTables.contains(tableToUnpin) {
            pinnedTables.removeAll(where: { $0 == tableToUnpin })
```

`pinnedTables.removeAll` would remove the tableToUnpin from pinnedTables, which is a copy of pinnedTablesDatabaseDictionary[hostName]?[databaseName]. The pinnedTablesDatabaseDictionary is unchanged. 

@Kaspik Please go through the most recent commit, I've reverted addToPinnedTablesDatabaseDictionary function back to how it was due to same issue as above. I don't see a clean way to dealing with local variables being copy of the `pinnedTablesDatabaseDictionary` collection, not a pointer/reference. Any suggestions welcome 😁

Kaspik commented:
There is no an easy way to do it as you shouldn't do it - pointer in Swift might mean that the object was mutated during that which is a wrong way to do it. Local variable should be mutated, worked on, and then reassigned back. I guess for now this is okay.

ShashwatC commented:
@Kaspik Got it, I think this should do the job-
```swift
        var pinnedTablesForHost = pinnedTablesDatabaseDictionary[hostName] ?? [:]
        var pinnedTables = pinnedTablesDatabaseDictionary[hostName]?[databaseName] ?? []
        pinnedTables.append(tableToPin)
        pinnedTablesForHost[databaseName] = pinnedTables
        pinnedTablesDatabaseDictionary[hostName] = pinnedTablesForHost
```

Kaspik commented:
You can leave it as it is, but:

```
var pinnedTablesForHost = pinnedTablesDatabaseDictionary[hostName] ?? [:]
var pinnedTables = pinnedTablesForHost[databaseName] ?? []
pinnedTables.append(tableToPin)
pinnedTablesForHost[databaseName] = pinnedTables
pinnedTablesDatabaseDictionary[hostName] = pinnedTablesForHost
```
would probably do that, yes. :)

ShashwatC commented:
I'll leave it as it is then. Don't think there would be any significant performance difference, and I find it easier to see what the function is doing in the way code is currently written (might be biased 😁 ), particularly the below part, which is somewhat obfuscated above. 
```swift
pinnedTablesDatabaseDictionary[hostName]?[databaseName]?.append(tableToPin)
```

Kaspik commented:
Sure, original comment was about `!` but as you switched to optionals, it works now. :)
